common/containers/Tabs/SendTransaction/index.jsx

```diff
@@ -404,36 +405,65 @@ export class SendTransaction extends React.Component {
     this.setState({ gasLimit: value, gasChanged: true });
   };
 
+  handleEverythingAmountChange = (value: string, unit: string): string => {
+    if (unit === 'ether') {
+      const { balance, gasPrice } = this.props;
+      const { gasLimit } = this.state;
+      const weiBalance = balance.toWei();
+      const bigGasLimit = new Big(gasLimit);
+      value = getBalanceMinusGasCosts(
+        bigGasLimit,
+        gasPrice,
+        weiBalance
+      ).amount.toString();
+    } else {
+      const tokenBalance = this.props.tokenBalances.find(
+        tokenBalance => tokenBalance.symbol === unit
+      );
+      if (!tokenBalance) {
+        throw new Error(`${unit}: not found in token balances;`);
+      }
+      value = tokenBalance.balance.toString();
+    }
+    return value;
+  };
+
   onAmountChange = (value: string, unit: string) => {
     if (value === 'everything') {
-      if (unit === 'ether') {
-        const { balance, gasPrice } = this.props;
-        const { gasLimit } = this.state;
-        const weiBalance = toWei(balance, 'ether');
-        value = getBalanceMinusGasCosts(
-          new Big(gasLimit),
-          new Big(gasPrice),
-          weiBalance
-        );
-      } else {
-        const tokenBalance = this.props.tokenBalances.find(
-          tokenBalance => tokenBalance.symbol === unit
-        );
-        if (!tokenBalance) {
-          return;
-        }
-        value = tokenBalance.balance.toString();
-      }
+      value = this.handleEverythingAmountChange(value, unit);
+    }
+    let transaction = this.state.transaction;
+    let generateDisabled = this.state.generateDisabled;
+    if (unit && unit !== this.state.unit) {
+      value = '';
+      transaction = null;
+      generateDisabled = true;
     }
     let token = this.props.tokens.find(x => x.symbol === unit);
     this.setState({
       value,
       unit,
-      token
+      token,
+      transaction,
+      generateDisabled
+    });
+  };
+
+  resetJustTx = async (): Promise<*> => {
```

Comment:
This could be cleaned up using arrow functions implict returns 
```js
resetJustTx = (): Promise<*> =>
  new Promise(resolve =>
    this.setState(
      {
        transaction: null
      },
      resolve
    )
  );
```

Reply:
👍 I suppose so! Kinda funky though, I haven't really seen this pattern before (but I suppose why not just have resolve be the callback if all you do is call it in the callback anyway).

Reply:
If we end up using async set state a lot, a HOF like this would come handy. (Pulled from a react-semantic-ui component which is why you see value as the second param)
```js 
		handleChange = itemToSet => (event, { value }) =>
			new Promise(resolve =>
				this.setState(prevState => {
					console.log('Previous state:', prevState, '\nForm value set:', {
						[itemToSet]: value
					})
					return { [itemToSet]: value }
				}, resolve)
			)
```